### ISSUE TITLE
DS School - Config Test

### ISSUE THREAD

**Issue by yeomko22** [yeomko22/TIL#125]
### Config test
- ML model은 config를 주입 받아서 학습을 진행하도록 구현한다.
- 그리고 config는 yaml 파일로 정의한다.
- yaml 파일로부터 config 객체를 생성하는 함수와 이를 모델에 주입해주는 함수를 각각 만든다.
- 이들에 대한 테스트 코드를 짜서 config에 대한 테스트를 진행할 수 있다. 
- 만일 개발자가 잘못된 config를 주입할 경우 에러를 발생시켜 줄 수 있도록 안전장치를 마련하는 것이다.
- 이 떄 pydantic 라이브러리를 유용하게 사용할 수 있다.

~~~python
def test_fetch_config_structure(tmpdir):
    # Given
    # We make use of the pytest built-in tmpdir fixture
    configs_dir = Path(tmpdir)
    config_1 = configs_dir / "sample_config.yml"
    config_1.write_text(TEST_CONFIG_TEXT)
    parsed_config = fetch_config_from_yaml(cfg_path=config_1)

    # When
    config = create_and_validate_config(parsed_config=parsed_config)

    # Then
    assert config.model_config
    assert config.app_config


def test_config_validation_raises_error_for_invalid_config(tmpdir):
    # Given
    # We make use of the pytest built-in tmpdir fixture
    configs_dir = Path(tmpdir)
    config_1 = configs_dir / "sample_config.yml"

    # invalid config attempts to set a prohibited loss
    # function which we validate against an allowed set of
    # loss function parameters.
    config_1.write_text(INVALID_TEST_CONFIG_TEXT)
    parsed_config = fetch_config_from_yaml(cfg_path=config_1)

    # When
    with pytest.raises(ValidationError) as excinfo:
        create_and_validate_config(parsed_config=parsed_config)

    # Then
    assert "not in the allowed set" in str(excinfo.value)
~~~

**Comment by yeomko22** [comment]
### test input data
- input data의 schema도 미리 정의해놓은 다음, validation test를 수행한다.
~~~python
def test_pipeline_drops_unnecessary_features(pipeline_inputs):
    # Given
    X_train, X_test, y_train, y_test = pipeline_inputs
    assert config.model_config.drop_features in X_train.columns

    # When
    # We use the scikit-learn Pipeline private method `_fit` which is called
    # by the `fit` method, since this allows us to access the transformed
    # dataframe. For other models we could use the `transform` method, but
    # the GradientBoostingRegressor does not have a `transform` method.
    X_transformed, _ = pipeline.price_pipe._fit(X_train, y_train)

    # Then
    assert config.model_config.drop_features in X_train.columns
    assert config.model_config.drop_features not in X_transformed.columns


def test_pipeline_transforms_temporal_features(pipeline_inputs):
    # Given
    X_train, X_test, y_train, y_test = pipeline_inputs

    # When
    # We use the scikit-learn Pipeline private method `_fit` which is called
    # by the `fit` method, since this allows us to access the transformed
    # dataframe. For other models we could use the `transform` method, but
    # the GradientBoostingRegressor does not have a `transform` method.
    X_transformed, _ = pipeline.price_pipe._fit(X_train, y_train)

    # Then
    assert (
        X_transformed.iloc[0]["YearRemodAdd"]
        == X_train.iloc[0]["YrSold"] - X_train.iloc[0]["YearRemodAdd"]
    )

~~~
